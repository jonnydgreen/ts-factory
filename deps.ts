// Path
import { join, resolve } from 'https://deno.land/std@0.147.0/path/mod.ts';
export const Path = {
  join,
  resolve,
};

// TS
export { ts } from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
export * as TSM from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
