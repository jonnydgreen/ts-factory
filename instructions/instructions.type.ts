import { Definition } from '../definitions/definitions.ts';

export type OpaqueString<TString extends string> = string & { __opaque: TString };

export type ItemOrArray<TItem> = TItem | TItem[];

export type Path = OpaqueString<'Path'>;

export enum InstructionType {
  SET = 'SET',
  ADD = 'ADD',
  INSERT = 'INSERT',
  REPLACE = 'REPLACE',
  UNSET = 'UNSET',
  REMOVE = 'REMOVE',
}

// TODO: review descriptions

/**
 * Add field to an array of fields on a parent node. E.g. statements
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the add operator associated with the defined node and field
 *  - Compile definition and apply to add operator
 *  - Move on to the next instruction
 */
export interface AddInstruction {
  type: InstructionType.ADD;
  field: string;
  path?: Path;
  definition: Definition;
}

/**
 * Set a field on a parent node. E.g. async keyword
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the set operator associated with the defined node and field
 *  - Compile definition and apply to set operator
 *  - Move on to the next instruction
 */
export interface SetInstruction {
  type: InstructionType.SET;
  path: Path;
  field: string;
  definition: Definition;
}

/**
 * Insert a field at a specific index within an array of fields on a parent node. E.g. statement at position 2
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the insert operator associated with the defined node and field
 *  - Ensure the array is of length at least the set position
 *  - Compile definition and apply to insert operator
 *  - Move on to the next instruction
 */
export interface InsertInstruction {
  type: InstructionType.INSERT;
  path: Path;
  field: string;
  index: number;
  definition: Definition;
}

/**
 * Replace a field at a specific index within an array of fields on a parent node. E.g. new statement at position 1
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the replace operator associated with the defined node and field
 *  - Ensure the array is of length at least the set position
 *  - Compile definition and apply to insert operator
 *  - Move on to the next instruction
 */
export interface ReplaceInstruction {
  type: InstructionType.REPLACE;
  path: Path;
  field: string;
  index: number;
  definition: Definition;
}

/**
 * Remove a field at a specific index within an array of fields on a parent node. E.g. remove statement at position 1
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the replace operator associated with the defined node and field
 *  - Ensure the array is of length at least the set position
 *  - Compile definition and apply to insert operator
 *  - Move on to the next instruction
 */
export interface RemoveInstruction {
  type: InstructionType.REMOVE;
  path: Path;
  field: string;
  index: number;
}

/**
 * Unset a field on node. E.g. type on a function declaration
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the replace operator associated with the defined node and field
 *  - Ensure the array is of length at least the set position
 *  - Compile definition and apply to insert operator
 *  - Move on to the next instruction
 */
export interface UnsetInstruction {
  type: InstructionType.UNSET;
  path: Path;
  field: string;
}

export type Instruction =
  | AddInstruction
  | SetInstruction
  | InsertInstruction
  | ReplaceInstruction
  | UnsetInstruction
  | RemoveInstruction;

export interface InstructionRule {
  instruction: InstructionType;
  condition: string;
  index?: number | string;
  field?: string;
}

export interface Instructions {
  id?: string;
  rules?: InstructionRule[];
}
