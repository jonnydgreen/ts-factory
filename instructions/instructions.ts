import {
  Instruction,
  InstructionType,
  NodeID,
  OpaqueString,
  SetInstruction,
} from './instructions.type.ts';
import { Definition } from '../definitions/definitions.ts';
import { jsonata, ts, tsm } from '../deps.ts';
import { Maybe } from '../types.ts';
import { StringUtils } from '../utils/utils.string.ts';
import { FunctionDeclarationInput } from '../definitions/function-declaration/function-declaration.type.ts';

export function assertDefinitionKind(
  definition: Definition,
  kind: tsm.ts.SyntaxKind,
): asserts definition is Extract<Definition, { kind: typeof kind }> {
  if (definition.kind !== kind) {
    throw new TypeError(
      `Definition of kind '${
        tsm.ts.SyntaxKind[definition.kind]
      }' does not match expected Node kind '${tsm.SyntaxKind[kind]}'`,
    );
  }
}

export function getDefinitionEntries(
  definition: Definition,
): [string, ItemOrArray<Definition>][] {
  const bannedKeys = ['kind', '__instructions'];
  return Object.entries(definition).filter(([key]) => !bannedKeys.includes(key));
}

export type ItemOrArray<TItem> = TItem | TItem[];

export function getFunctionDeclarationNodeByDefinitionKey(
  node: tsm.FunctionDeclaration,
  fieldName: string,
): Maybe<ItemOrArray<tsm.Node>> {
  switch (fieldName as keyof FunctionDeclarationInput) {
    case 'name': {
      return node.getNameNode();
    }
    case 'type': {
      return node.getReturnTypeNode();
    }
    case 'parameters': {
      return node.getParameters();
    }
  }
}

export function getFieldNodeByDefinitionKey(
  node: tsm.Node,
  fieldName: string,
): Maybe<ItemOrArray<tsm.Node>> {
  switch (node.getKind()) {
    case ts.SyntaxKind.FunctionDeclaration: {
      return getFunctionDeclarationNodeByDefinitionKey(
        node as tsm.FunctionDeclaration,
        fieldName,
      );
    }
  }
  // TODO: we should probably be less clever here
  // and provide a mapping for each Syntax Kind
  const getFunctionName = `get${StringUtils.upperFirst(fieldName)}` as keyof tsm.Node;
  const getFunction = node[getFunctionName] as () => Maybe<ItemOrArray<tsm.Node>>;
  if (typeof getFunction !== 'function') {
    throw new TypeError(
      `Unable to get field of name '${fieldName}' from node of kind '${node.getKindName()}'`,
    );
  }
  return getFunction.call(node);
}

export function compileDefaultNodeArrayInstructions(
  nodeID: NodeID,
  definitions: Definition[],
  field: string,
  instructionType?: InstructionType,
  index?: number,
): Instruction[] {
  return definitions.map(({ __instructions, ...definitionItem }) => {
    if (instructionType === InstructionType.INSERT) {
      return {
        type: instructionType,
        definition: definitionItem,
        field,
        nodeID,
        // TODO: make this better
        index: index as number,
      };
    }
    return {
      type: instructionType ?? InstructionType.ADD,
      definition: definitionItem,
      field,
      nodeID,
    };
  });
}

export function compileDefaultNodeInstructions(
  nodeID: NodeID,
  definition: Definition,
  field: string,
  instructionType?: InstructionType.SET,
): SetInstruction[] {
  // TODO: assert the correct instruction type somewhere
  // Don't want any casts if we can help it
  return [{
    type: instructionType ?? InstructionType.SET,
    definition,
    field,
    nodeID,
  }];
}

export function compileInstructions(
  nodeID: NodeID,
  definitionOrDefinitions: ItemOrArray<Definition>,
  field: string,
  instructionType?: InstructionType,
  index?: number,
): Instruction[] {
  // If an array, we compile an ADD instruction for each definition.
  if (Array.isArray(definitionOrDefinitions)) {
    return compileDefaultNodeArrayInstructions(
      nodeID,
      definitionOrDefinitions,
      field,
      instructionType,
      index,
    );
  }
  // Otherwise, we use a set instruction using the defined values
  const { __instructions, ...definition } = definitionOrDefinitions;
  return compileDefaultNodeInstructions(
    nodeID,
    definition,
    field,
    instructionType as InstructionType.SET,
  );
}

export function isDefined<T>(nodeOrNodes: Maybe<ItemOrArray<T>>): boolean {
  if (!nodeOrNodes) {
    return false;
  }

  if (Array.isArray(nodeOrNodes) && nodeOrNodes.length === 0) {
    return false;
  }

  return true;
}

export function createOpaqueString<T extends OpaqueString<string>>(input: string): T {
  return input as T;
}

export function createNodeID(id = '', ...nextIDs: string[]): NodeID {
  let newId = id;
  for (const nextID of nextIDs) {
    if (Number.isInteger(Number(nextID))) {
      newId += `[${nextID}]`;
    } else {
      newId += `.${nextID}`;
    }
  }
  return createOpaqueString<NodeID>(newId.replace(/^\.+/, ''));
}

export function isDefinition(definition: unknown): definition is Definition {
  return typeof definition === 'object' && definition !== null &&
    Number.isInteger((definition as Definition).kind);
}

export function generateInstructions(
  node: tsm.Node,
  definition: Definition,
  nodeID: NodeID = createNodeID(),
): Instruction[] {
  assertDefinitionKind(definition, node.getKind());

  const instructions: Instruction[] = [];
  for (
    const [fieldName, fieldDefinitionOrDefinitions] of getDefinitionEntries(definition)
  ) {
    // Get the running node field
    const nodeValueOrValues = getFieldNodeByDefinitionKey(node, fieldName);

    // If no node or nodes found, we can immediately build an instruction
    if (!isDefined(nodeValueOrValues)) {
      instructions.push(
        ...compileInstructions(nodeID, fieldDefinitionOrDefinitions, fieldName),
      );
      continue;
    }

    // Otherwise, we have found a node

    // If the fieldDefinition is an array, we loop through these and process in turn
    if (Array.isArray(fieldDefinitionOrDefinitions)) {
      for (const fieldDefinition of fieldDefinitionOrDefinitions) {
        // If no instructions, we compile using the default instruction rules
        if (!fieldDefinition.__instructions) {
          instructions.push(
            ...compileDefaultNodeArrayInstructions(nodeID, [fieldDefinition], fieldName),
          );
          // TODO: we need to decide what to do after ^
        } else if (fieldDefinition.__instructions.id) {
          const compiledQuery = jsonata(fieldDefinition.__instructions.id);
          // TODO: assert here
          const nodes = (nodeValueOrValues as tsm.Node[]);
          const foundNodeIndex = nodes.findIndex((node) =>
            !!compiledQuery.evaluate(node.compilerNode)
          );
          if (foundNodeIndex === -1) {
            instructions.push(
              ...compileInstructions(nodeID, [fieldDefinition], fieldName),
            );
          } else {
            const nodeId = createNodeID(nodeID, fieldName, foundNodeIndex.toString());
            if (isDefinition(fieldDefinition)) {
              instructions.push(
                ...generateInstructions(nodes[foundNodeIndex], fieldDefinition, nodeId),
              );
            }
          }
          // If rules are defined, we process accordingly
          // At the moment, we do not process ID and rules at the same time
          // because the behaviour is not yet defined.
        }

        if (fieldDefinition.__instructions?.rules) {
          for (const rule of fieldDefinition.__instructions.rules) {
            const compiledQuery = jsonata(rule.condition);
            const nodes = (nodeValueOrValues as tsm.Node[]);
            const ruleMatch = nodes.some((node) =>
              !!compiledQuery.evaluate(node.compilerNode)
            );
            let index: Maybe<number>;
            if (typeof rule.index === 'string') {
              const result = Number(jsonata(rule.index).evaluate(nodes));
              if (!Number.isInteger(result) || result >= nodes.length) {
                throw new TypeError(
                  `Invalid result, must be integer within the array length: ${result}`,
                );
              }
              index = result;
            } else {
              index = rule.index;
            }
            // If a rule matches we process the results
            if (ruleMatch) {
              instructions.push(
                ...compileInstructions(
                  nodeID,
                  [fieldDefinition],
                  fieldName,
                  rule.instruction,
                  index,
                ),
              );
            }
          }
        }
      }
    } else {
      const fieldDefinition = fieldDefinitionOrDefinitions;
      if (fieldDefinition.__instructions?.id) {
        const compiledQuery = jsonata(fieldDefinition.__instructions.id);
        const foundNode = [nodeValueOrValues as tsm.Node].find((node) =>
          compiledQuery.evaluate(node.compilerNode)
        );
        if (!foundNode) {
          instructions.push(
            ...compileInstructions(nodeID, fieldDefinition, fieldName),
          );
        } else {
          const nodeId = createNodeID(nodeID, fieldName);
          if (isDefinition(fieldDefinition)) {
            instructions.push(
              ...generateInstructions(foundNode, fieldDefinition, nodeId),
            );
          }
        }
      }
    }
  }
  return instructions;
}

// export function findNode(
//   node: tsm.Node,
//   fieldName: string,
//   fieldDefinition: Definition,
// ): Maybe<tsm.Node | tsm.Node[]> {
//   if (!fieldDefinition.__instructions) {
//     const getFieldFn =
//       node[`get${StringUtils.upperFirst(fieldName)}` as keyof typeof node];
//     if (typeof getFieldFn === 'function') {
//       return (getFieldFn as () => Maybe<tsm.Node | tsm.Node[]>)();
//     }
//     return undefined;
//   }

//   return undefined;
// }
