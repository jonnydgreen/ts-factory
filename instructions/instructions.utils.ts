import { ts, tsm } from '../deps.ts';
import { Maybe } from '../types.ts';
import { ItemOrArray } from './instructions.type.ts';
import { StringUtils } from '../utils/utils.string.ts';
import { FunctionDeclarationInput } from '../definitions/function-declaration/function-declaration.type.ts';
import { Definition } from '../definitions/definitions.ts';

export function getFunctionDeclarationNodeByDefinitionKey(
  node: tsm.FunctionDeclaration,
  fieldName: string,
): Maybe<ItemOrArray<tsm.Node>> {
  switch (fieldName as keyof FunctionDeclarationInput) {
    case 'name': {
      return node.getNameNode();
    }
    case 'type': {
      return node.getReturnTypeNode();
    }
    case 'parameters': {
      return node.getParameters();
    }
    case 'modifiers': {
      return node.getModifiers();
    }
    case 'body': {
      return node.getBody();
    }
  }
}

// TODO: consider a new name for this
export function getFieldNodeByDefinitionKey(
  node: tsm.Node,
  fieldName: string,
): Maybe<ItemOrArray<tsm.Node>> {
  switch (node.getKind()) {
    case ts.SyntaxKind.FunctionDeclaration: {
      return getFunctionDeclarationNodeByDefinitionKey(
        node as tsm.FunctionDeclaration,
        fieldName,
      );
    }
  }
  // TODO: we should probably be less clever here
  // and provide a mapping for each Syntax Kind
  const getFunctionName = `get${StringUtils.upperFirst(fieldName)}` as keyof tsm.Node;
  const getFunction = node[getFunctionName] as () => Maybe<ItemOrArray<tsm.Node>>;
  if (typeof getFunction !== 'function') {
    throw new TypeError(
      `Unable to get field of name '${fieldName}' from node of kind '${node.getKindName()}'`,
    );
  }
  return getFunction.call(node);
}

export function getNodeByPath(currentNode: tsm.Node, path = ''): tsm.Node {
  if (path === '') {
    return currentNode;
  }
  const [location, ...remainingPath] = path
    .replaceAll('[', '.')
    .replaceAll(']', '.')
    .replace(/\.+/g, '.')
    .split('.');
  const nextNodeOrNodes = getFieldNodeByDefinitionKey(currentNode, location);
  if (Array.isArray(nextNodeOrNodes)) {
    const index = Number(remainingPath?.[0]);
    if (!Number.isInteger(index)) {
      throw new TypeError(
        `Next node is an array of nodes but the following ID is not an array index '${remainingPath
          ?.[0]}'`,
      );
    }
    const nextNode = nextNodeOrNodes[index];
    if (typeof nextNode === 'undefined') {
      throw new TypeError(`Node not found at index '${index}'`);
    }
    const [, ...nextIdParts] = remainingPath;
    return getNodeByPath(nextNode, nextIdParts.join('.'));
  }

  console.log(currentNode.getKindName(), location);

  if (typeof nextNodeOrNodes === 'undefined') {
    throw new TypeError(
      `Node not found at location '${location}' for Path '${path}'`,
    );
  }

  if (remainingPath.length === 0) {
    return nextNodeOrNodes as tsm.Node;
  } else {
    return getNodeByPath(nextNodeOrNodes as tsm.Node, remainingPath.join('.'));
  }
}

export function shouldBuildNodeFromDefinition<TNode extends ts.Node>(
  definition?: Definition,
): Maybe<TNode> {
  if (typeof definition === 'undefined') {
    return;
  }
  return buildNodeFromDefinition<TNode>(definition);
}

// TODO: any
export function buildNodeFromDefinition<TReturn extends ts.Node>(
  definition: Definition,
): TReturn {
  let node: ts.Node;

  switch (definition.kind) {
    // TODO: uncomment
    // case ts.SyntaxKind.ReturnStatement: {
    //   const expression = shouldBuildNode(definition.expression);
    //   node = ts.factory.createReturnStatement(expression);
    //   break;
    // }
    // case ts.SyntaxKind.CallExpression: {
    //   const expression = buildNode(definition.expression);
    //   node = ts.factory.createCallExpression(
    //     expression,
    //     definition.typeArguments?.map(buildNode),
    //     definition.arguments?.map(buildNode),
    //   );
    //   break;
    // }
    // case ts.SyntaxKind.PropertyAccessExpression: {
    //   const expression = buildNode(definition.expression);
    //   const name = buildNode(definition.name);
    //   node = ts.factory.createPropertyAccessExpression(expression, name);
    //   break;
    // }
    // case ts.SyntaxKind.StringLiteral: {
    //   node = ts.factory.createStringLiteral(definition.text);
    //   break;
    // }
    case ts.SyntaxKind.Identifier: {
      node = ts.factory.createIdentifier(definition.text);
      break;
    }
    // Modifier
    case ts.SyntaxKind.AsyncKeyword:
    case ts.SyntaxKind.ExportKeyword: {
      node = ts.factory.createModifier(definition.kind);
      break;
    }
    case ts.SyntaxKind.Block: {
      node = ts.factory.createBlock(
        definition.statements.map<ts.Statement>(buildNodeFromDefinition),
        definition.multiline,
      );
      break;
    }
    case ts.SyntaxKind.FunctionDeclaration: {
      node = ts.factory.createFunctionDeclaration(
        undefined,
        definition.modifiers?.map<ts.Modifier>(buildNodeFromDefinition),
        undefined,
        typeof definition.name === 'string'
          ? definition.name
          : shouldBuildNodeFromDefinition<ts.Identifier>(definition.name),
        undefined,
        definition.parameters as ts.ParameterDeclaration[],
        shouldBuildNodeFromDefinition<ts.TypeNode>(definition.type),
        shouldBuildNodeFromDefinition<ts.Block>(definition.body),
      );
      break;
    }
    default: {
      throw new TypeError(
        `TS Node Kind not supported: '${definition.kind}|${
          ts.SyntaxKind[definition.kind]
        }'`,
      );
    }
  }

  // TODO: uncomment
  // if (definition.leadingTrivia) {
  //   ts.addSyntheticLeadingComment(
  //     node,
  //     definition.leadingTrivia.kind,
  //     definition.leadingTrivia.text,
  //     definition.leadingTrivia.hasTrailingNewLine,
  //   );
  // }

  return node as TReturn;
}

export function processFunctionDeclarationFieldNode(
  parentNode: tsm.FunctionDeclaration,
  field: string,
  nodeToAdd: ts.Node,
): void {
  switch (field as keyof FunctionDeclarationInput) {
    case 'modifiers': {
      const kind = (nodeToAdd as ts.Modifier).kind;
      switch (kind) {
        case ts.SyntaxKind.ExportKeyword: {
          parentNode.setIsExported(true);
          break;
        }
      }
      // TODO: assertNever
      break;
    }
      // // TODO: assertNever
      // default: {
      // }
  }
}

export function defaultAddNodeToField(
  parentNode: tsm.Node,
  field: string,
  rawNode: string,
): void {
  const addFunctionName = `add${StringUtils.upperFirst(field)}` as keyof tsm.Node;
  const addFunction = parentNode[addFunctionName] as (input: string[]) => unknown;
  if (typeof addFunction !== 'function') {
    throw new TypeError(
      `Unable to add node to field of name '${field}' for node of kind '${parentNode.getKindName()}'`,
    );
  }
  addFunction.call(parentNode, [rawNode]);
}

export function addNodeToField(
  parentNode: tsm.Node,
  field: string,
  nodeToAdd: ts.Node,
): void {
  const rawNode = tsm.printNode(nodeToAdd, { removeComments: false });

  switch (parentNode.getKind()) {
    case ts.SyntaxKind.FunctionDeclaration: {
      return processFunctionDeclarationFieldNode(
        parentNode as tsm.FunctionDeclaration,
        field,
        nodeToAdd,
      );
    }

    default: {
      defaultAddNodeToField(parentNode, field, rawNode);
    }
  }
}
