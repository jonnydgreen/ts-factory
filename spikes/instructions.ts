import { ts, TSM as tsm } from '../deps.ts';
import { convertToNodeFn, getNodeById } from './get-node-by-id.ts';
import { Definition } from './types.ts';

const { SyntaxKind, factory } = ts;

export enum InstructionType {
  SET,
  ADD,
  INSERT,
  REPLACE,
}

/**
 * Add field to an array of fields on a parent node. E.g. statements
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the add operator associated with the defined node and field
 *  - Compile definition and apply to add operator
 *  - Move on to the next instruction
 */
export interface AddInstruction {
  type: InstructionType.ADD;
  field: string;
  nodeId?: string;
  definition: Definition;
}

/**
 * Set a field on a parent node. E.g. async keyword
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the set operator associated with the defined node and field
 *  - Compile definition and apply to set operator
 *  - Move on to the next instruction
 */
export interface SetInstruction {
  type: InstructionType.SET;
  nodeId: string;
  field: string;
  definition: Definition;
}

/**
 * Insert a field at a specific index within an array of fields on a parent node. E.g. statement at position 2
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the insert operator associated with the defined node and field
 *  - Ensure the array is of length at least the set position
 *  - Compile definition and apply to insert operator
 *  - Move on to the next instruction
 */
export interface InsertInstruction {
  type: InstructionType.INSERT;
}

/**
 * Replace a field at a specific index within an array of fields on a parent node. E.g. new statement at position 1
 *
 * Flow:
 *  - Get Node by ID
 *  - Infer the kind of the Node
 *  - Find the replace operator associated with the defined node and field
 *  - Ensure the array is of length at least the set position
 *  - Compile definition and apply to insert operator
 *  - Move on to the next instruction
 */
export interface ReplaceInstruction {
  type: InstructionType.REPLACE;
}

export type Instruction =
  | AddInstruction
  | SetInstruction
  | InsertInstruction
  | ReplaceInstruction;

export function assertNever(input: never): never {
  throw new Error(`Should never receive input '${input}'`);
}

export function shouldBuildNode(ast: any): any {
  if (typeof ast === 'undefined') {
    return;
  }
  return buildNode(ast);
}

factory.createCallExpression;

export function buildNode(ast: any): any {
  let node: ts.Node;
  switch (ast?.kind) {
    case SyntaxKind.ReturnStatement: {
      const expression = shouldBuildNode(ast.expression);
      node = factory.createReturnStatement(expression);
      break;
    }
    case SyntaxKind.CallExpression: {
      const expression = buildNode(ast.expression);
      node = factory.createCallExpression(
        expression,
        ast.typeArguments?.map(buildNode),
        ast.arguments?.map(buildNode),
      );
      break;
    }
    case SyntaxKind.PropertyAccessExpression: {
      const expression = buildNode(ast.expression);
      const name = buildNode(ast.name);
      node = factory.createPropertyAccessExpression(expression, name);
      break;
    }
    case SyntaxKind.Identifier: {
      node = factory.createIdentifier(ast.text);
      break;
    }
    case SyntaxKind.StringLiteral: {
      node = factory.createStringLiteral(ast.text);
      break;
    }
    case SyntaxKind.AnyKeyword:
    case SyntaxKind.BigIntKeyword:
    case SyntaxKind.BooleanKeyword:
    case SyntaxKind.IntrinsicKeyword:
    case SyntaxKind.NeverKeyword:
    case SyntaxKind.NumberKeyword:
    case SyntaxKind.ObjectKeyword:
    case SyntaxKind.StringKeyword:
    case SyntaxKind.SymbolKeyword:
    case SyntaxKind.UndefinedKeyword:
    case SyntaxKind.UnknownKeyword:
    case SyntaxKind.VoidKeyword: {
      node = factory.createKeywordTypeNode(ast.kind);
      break;
    }
    default: {
      throw new Error(
        `TS Node Kind not supported: '${ts.SyntaxKind[ast.kind]}'`,
      );
    }
  }

  if (ast.leadingTrivia) {
    ts.addSyntheticLeadingComment(
      node,
      ast.leadingTrivia.kind,
      ast.leadingTrivia.text,
      ast.leadingTrivia.hasTrailingNewLine,
    );
  }

  return node;
}

export function processInstruction(
  sourceFile: tsm.SourceFile,
  instruction: Instruction,
) {
  switch (instruction.type) {
    case InstructionType.ADD: {
      const node = getNodeById(sourceFile, instruction.nodeId ?? '');
      const fieldNodesToAdd = [buildNode(instruction.definition)];
      const functionName = convertToNodeFn('add', instruction.field);
      const addNodeFieldsFunction = (
        node[functionName as keyof typeof node] as (input: unknown[]) => void
      ).bind(node);
      addNodeFieldsFunction(
        fieldNodesToAdd.map((node) => tsm.printNode(node, { removeComments: false })),
      );
      break;
    }
    case InstructionType.SET: {
      const node = getNodeById(sourceFile, instruction.nodeId);
      const fieldNodeToSet = buildNode(instruction.definition);
      const functionName = convertToNodeFn('set', instruction.field);
      console.log({ functionName, node });
      const setNodeFieldFunction = (
        node[functionName as keyof typeof node] as (input: unknown) => void
      ).bind(node);
      setNodeFieldFunction(
        tsm.printNode(fieldNodeToSet, { removeComments: false }),
      );
      break;
    }
    case InstructionType.REPLACE:
    case InstructionType.INSERT: {
      throw new Error(`Instruction type ${instruction.type} not supported.`);
    }
    default: {
      assertNever(instruction);
    }
  }
}
