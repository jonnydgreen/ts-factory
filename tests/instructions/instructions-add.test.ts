import { ts, tsm } from '../../deps.ts';
import { generateInstructions } from '../../instructions/instructions.ts';
import { assertSnapshot, blocks } from '../../test.deps.ts';
import { createTestName, sanitiseInstructions, TestDefinition } from '../test-utils.ts';

blocks.describe('Instructions', () => {
  blocks.describe(`Generate ADD Instruction`, () => {
    const definitions: TestDefinition[] = [
      {
        name: createTestName(
          'should generate an ADD Instruction if',
          'the field is an array of nodes',
          'the field nodes don\'t exist',
          'no instructions are specified on the field definition',
        ),
        input: {
          kind: ts.SyntaxKind.SourceFile,
          statements: [
            {
              kind: ts.SyntaxKind.FunctionDeclaration,
              parameters: [],
            },
          ],
        },
      },
      {
        name: createTestName(
          'should generate an ADD Instruction if',
          'the field is an array of nodes',
          'the field nodes don\'t exist',
          'instructions are specified on the field definition',
        ),
        input: {
          kind: ts.SyntaxKind.SourceFile,
          statements: [
            {
              __instructions: {
                id: 'some-instruction',
                rules: [],
              },
              kind: ts.SyntaxKind.FunctionDeclaration,
              parameters: [],
            },
          ],
        },
      },
      {
        name: createTestName(
          'should generate an ADD Instruction if',
          'the field is an array of nodes',
          'the field nodes exist',
          'no instructions are specified on the field definition',
        ),
        input: {
          kind: ts.SyntaxKind.SourceFile,
          statements: [
            {
              kind: ts.SyntaxKind.FunctionDeclaration,
              parameters: [],
            },
          ],
        },
        sourceFileContents: `
          export function hello() {}
        `,
      },
      {
        name: createTestName(
          'should generate an ADD Instruction if',
          'the field is an array of nodes',
          'the field node is not found using the defined instruction ID',
        ),
        input: {
          kind: ts.SyntaxKind.SourceFile,
          statements: [
            {
              __instructions: {
                // Looks like there is an AST
                id: 'name.escapedText="hello"',
              },
              kind: ts.SyntaxKind.FunctionDeclaration,
              name: {
                kind: ts.SyntaxKind.Identifier,
                // TODO: can we use bindings here?
                text: 'hello',
              },
              parameters: [],
            },
          ],
        },
        sourceFileContents: `
          export function foo() {}
        `,
      },
      {
        name: createTestName(
          'should not generate an ADD Instruction if',
          'the field is an array of nodes',
          'the field node is found using the defined instruction ID',
        ),
        input: {
          kind: ts.SyntaxKind.SourceFile,
          statements: [
            {
              __instructions: {
                // TODO: Looks like there is an AST we can use instead
                id: 'name.text="hello"',
              },
              kind: ts.SyntaxKind.FunctionDeclaration,
              name: {
                kind: ts.SyntaxKind.Identifier,
                // TODO: can we use bindings here when we compile a template?
                text: 'hello',
              },
              parameters: [],
            },
          ],
        },
        sourceFileContents: `
          export function hello() {}
        `,
      },
      {
        name: createTestName(
          'should generate an ADD Instruction if',
          'the field is an array of nodes',
          'the field is a nested definition',
        ),
        input: {
          kind: ts.SyntaxKind.SourceFile,
          statements: [
            {
              __instructions: {
                id: 'name.text="hello"',
              },
              kind: ts.SyntaxKind.FunctionDeclaration,
              name: {
                kind: ts.SyntaxKind.Identifier,
                // TODO: can we use bindings here when we compile a template?
                text: 'hello',
              },
              modifiers: [
                { kind: ts.SyntaxKind.ExportKeyword },
                { kind: ts.SyntaxKind.AsyncKeyword },
              ],
              parameters: [],
            },
          ],
        },
        sourceFileContents: `
          function hello() {}
        `,
      },
    ];

    for (const definition of definitions) {
      blocks.it(definition.name, async (t) => {
        // Arrange
        const project = new tsm.Project();
        const sourceFile = project.createSourceFile(
          `${crypto.randomUUID()}.ts`,
          definition.sourceFileContents,
        );

        // Act
        const result = generateInstructions(sourceFile, definition.input);

        // Assert
        await assertSnapshot(t, sanitiseInstructions(result));
      });
    }
  });
});
