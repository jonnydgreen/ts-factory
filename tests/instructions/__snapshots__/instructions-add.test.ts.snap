export const snapshot = {};

snapshot[`Instructions > Generate ADD Instruction > should generate an ADD Instruction if the field is an array of nodes, the field nodes don't exist and no instructions are specified on the field definition 1`] = `
[
  {
    definition: {
      body: {
        kind: "Block",
        statements: [
        ],
      },
      kind: "FunctionDeclaration",
      name: "hello",
      parameters: [
      ],
    },
    field: "statements",
    path: "",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > Generate ADD Instruction > should generate an ADD Instruction if the field is an array of nodes, the field nodes don't exist and instructions are specified on the field definition 1`] = `
[
  {
    definition: {
      body: {
        kind: "Block",
        statements: [
        ],
      },
      kind: "FunctionDeclaration",
      name: "hello",
      parameters: [
      ],
    },
    field: "statements",
    path: "",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > Generate ADD Instruction > should generate an ADD Instruction if the field is an array of nodes, the field nodes exist and no instructions are specified on the field definition 1`] = `
[
  {
    definition: {
      body: {
        kind: "Block",
        statements: [
        ],
      },
      kind: "FunctionDeclaration",
      name: "hello",
      parameters: [
      ],
    },
    field: "statements",
    path: "",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > Generate ADD Instruction > should generate an ADD Instruction if the field is an array of nodes and the field node is not found using the defined instruction ID 1`] = `
[
  {
    definition: {
      body: {
        kind: "Block",
        statements: [
        ],
      },
      kind: "FunctionDeclaration",
      name: {
        kind: "Identifier",
        text: "hello",
      },
      parameters: [
      ],
    },
    field: "statements",
    path: "",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > Generate ADD Instruction > should generate an ADD Instruction if the field is an array of nodes and the field is a nested definition 1`] = `
[
  {
    definition: {
      kind: "ExportKeyword",
    },
    field: "modifiers",
    path: "statements[0]",
    type: "ADD",
  },
  {
    definition: {
      kind: "AsyncKeyword",
    },
    field: "modifiers",
    path: "statements[0]",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > Generate ADD Instruction > should not generate an ADD Instruction if the field is an array of nodes and the field node is found using the defined instruction ID 1`] = `
[
]
`;

snapshot[`Instructions > Process ADD Instruction > should process an ADD Instruction if the field is an array of nodes, the field nodes don't exist and no instructions are specified on the field definition 1`] = `"function hello() { }"`;

snapshot[`Instructions > Process ADD Instruction > should process an ADD Instruction if the field is an array of nodes and the field is a nested definition 1`] = `"export default async function hello() { }"`;
